packages/shared/utils/types.ts

```diff
@@ -67,6 +67,8 @@ export type ShallowUnwrapRef<T> = T extends Ref<infer P> ? P : T
 
 export type Awaitable<T> = Promise<T> | T
 
+export type Thenable<T> = T & PromiseLike<T>
```

Comment:
Couldn't we just make `Awaitable<T> = PromiseLike<T> | T` instead of creating a new type? (their names do not indicates their difference either)

Reply:
union and intersection are different types. Anything of type `T` is of type `T | PromiseLike<T>` but might not be of type `T & PromiseLike<T>`.

Reply:
Ah I see. The reason it's called `Awaitable` instead of `MaybeAwaitable` because non-Promise values also awaitable.

For the union, I don't honestly think `Thenable` is a good name here. I don't have a good name in my mind - if we didn't come up with a distinguishable name in the end, I'd suggest we keep things as-is to not having this refactor.

Maybe `WithPromiseLike` or `WithPromiseAPI`?

Reply:
Fair. The motivation for the refactor is that sometimes I found it tedious to explicitly type the return, though most of time unnecessary. A shorter alias will make it eariser  for that and also reduces repeats in the source code. 

From the definition, the union of `T | Promise(Like)<T>` **is** `MaybePromiseLike`. I suggested the name `MaybeAwaitable` because that's already called `Awaitable`. Meanwhile, the intersection, `T & Promise(Like)<T>` **is** `Awaitable`, but that's taken by the union.

The function `const fn = () => 1`, which is of type `() => number`, is of course of type `() => number | Promise<number>`. I would not say `fn() = 1` is awaitable, though `JS` wraps it into a promise and `await fn()` is allowed.

I think `WithPromiseLike` is a good choice.

Reply:
@haoyun do you need any help/info to finish this up? :)

Reply:
Yeah I think `WithPromiseLike` is a good name